### ISSUE TITLE
[distributions] dirichlet pathwise gradient does not work well with .expand

### ISSUE THREAD

**Issue by fehiepsi** [pytorch/pytorch#9521]
The issue happens in pytorch master: when the expanded shape has a singleton (1) dimension. Samples get constant values at some dimension. This does not happen in pytorch 0.4.
Here is the code to replicate the issue
```
import torch
c = torch.tensor([1., 1.])
a = torch.distributions.dirichlet._Dirichlet.apply(c.expand(2, 1, 2))
b = torch.distributions.dirichlet._Dirichlet.apply(c.expand(2, 1, 2))
print(bool((a[1, 0, :] == b[1, 0, :]).all()))  # <- always True
```
@SsnL @fritzo

**Comment by alicanb** [comment]
It's not only dirichlet; `_standard_gamma` and `poisson` have this problem as well. I think it's due to the way they're implemented in `Distributions.cpp`

**Comment by SsnL** [comment]
I’ll take a look after 0.4.1

**Comment by neerajprad** [comment]
I think this should be closed now.
